### DISCUSSION TITLE
PowerDelta calculation lacks precision on small loads

### DISCUSSION THREAD

**Discussion by isorin** [arendst/Tasmota#11727] [Q&A] [answered]
### PROBLEM DESCRIPTION
Using "WattRes 1" so power is reported with 1 decimal precision.
and PowerDelta "15" which should trigger at 15% but looks like differences between 6.6W and 7.1W are enough to trigger it.
Looking at the logs it seems the delta calculations are done using only integers.

```
8:31:30.804 CMD: PowerDelta 15
18:31:30.813 MQT: stat/tasmota_0329F7/RESULT = {"PowerDelta1":15}
18:32:13.552 MQT: tele/tasmota_0329F7/MARGINS = {"MARGINS":{"PowerDelta":1}}
18:32:13.560 MQT: tele/tasmota_0329F7/SENSOR = {"Time":"2021-04-13T18:32:13","ENERGY":{"TotalStartTime":"2021-03-01T11:31:01","Total":16.33415,"Yesterday":0.40627,"Today":0.28024,"Power":7.1,"ApparentPower":15.6,"ReactivePower":13.9,"Factor":0.45,"Voltage":241,"Current":0.065}}
18:32:14.552 MQT: tele/tasmota_0329F7/MARGINS = {"MARGINS":{"PowerDelta":-1}}
18:32:14.561 MQT: tele/tasmota_0329F7/SENSOR = {"Time":"2021-04-13T18:32:14","ENERGY":{"TotalStartTime":"2021-03-01T11:31:01","Total":16.33415,"Yesterday":0.40627,"Today":0.28024,"Power":6.6,"ApparentPower":15.9,"ReactivePower":14.5,"Factor":0.42,"Voltage":241,"Current":0.066}}
18:32:43.549 MQT: tele/tasmota_0329F7/MARGINS = {"MARGINS":{"PowerDelta":1}}
18:32:43.557 MQT: tele/tasmota_0329F7/SENSOR = {"Time":"2021-04-13T18:32:43","ENERGY":{"TotalStartTime":"2021-03-01T11:31:01","Total":16.33420,"Yesterday":0.40627,"Today":0.28029,"Power":7.0,"ApparentPower":15.7,"ReactivePower":14.0,"Factor":0.45,"Voltage":241,"Current":0.065}}
18:32:44.551 MQT: tele/tasmota_0329F7/MARGINS = {"MARGINS":{"PowerDelta":-1}}
18:32:44.559 MQT: tele/tasmota_0329F7/SENSOR = {"Time":"2021-04-13T18:32:44","ENERGY":{"TotalStartTime":"2021-03-01T11:31:01","Total":16.33421,"Yesterday":0.40627,"Today":0.28030,"Power":6.7,"ApparentPower":15.7,"ReactivePower":14.2,"Factor":0.43,"Voltage":241,"Current":0.065}}
18:33:43.541 MQT: tele/tasmota_0329F7/MARGINS = {"MARGINS":{"PowerDelta":1}}
18:33:43.549 MQT: tele/tasmota_0329F7/SENSOR = {"Time":"2021-04-13T18:33:43","ENERGY":{"TotalStartTime":"2021-03-01T11:31:01","Total":16.33432,"Yesterday":0.40627,"Today":0.28041,"Power":7.0,"ApparentPower":15.6,"ReactivePower":14.0,"Factor":0.45,"Voltage":240,"Current":0.065}}
18:33:44.544 MQT: tele/tasmota_0329F7/MARGINS = {"MARGINS":{"PowerDelta":-1}}
18:33:44.552 MQT: tele/tasmota_0329F7/SENSOR = {"Time":"2021-04-13T18:33:44","ENERGY":{"TotalStartTime":"2021-03-01T11:31:01","Total":16.33432,"Yesterday":0.40627,"Today":0.28041,"Power":6.6,"ApparentPower":15.7,"ReactivePower":14.2,"Factor":0.42,"Voltage":241,"Current":0.065}}
18:34:13.544 MQT: tele/tasmota_0329F7/MARGINS = {"MARGINS":{"PowerDelta":1}}
18:34:13.551 MQT: tele/tasmota_0329F7/SENSOR = {"Time":"2021-04-13T18:34:13","ENERGY":{"TotalStartTime":"2021-03-01T11:31:01","Total":16.33437,"Yesterday":0.40627,"Today":0.28046,"Power":7.1,"ApparentPower":15.6,"ReactivePower":13.9,"Factor":0.45,"Voltage":241,"Current":0.065}}
18:34:14.543 MQT: tele/tasmota_0329F7/MARGINS = {"MARGINS":{"PowerDelta":-1}}
18:34:14.553 MQT: tele/tasmota_0329F7/SENSOR = {"Time":"2021-04-13T18:34:14","ENERGY":{"TotalStartTime":"2021-03-01T11:31:01","Total":16.33437,"Yesterday":0.40627,"Today":0.28046,"Power":6.8,"ApparentPower":15.9,"ReactivePower":14.4,"Factor":0.43,"Voltage":241,"Current":0.066}}
```

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [ ] Device used (e.g., Sonoff Basic): https://templates.blakadder.com/aoycocr_U3S.html
- [ ] Tasmota binary firmware version number used: Tasmota 9.3.1
  - [x] Pre-compiled
  - [ ] Self-compiled
- [ ] Flashing tools used: OTA upgrade
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
10:32:01.122 CMD: Backlog Template; Module; GPIO 255
10:32:01.190 MQT: stat/tasmota_0329F7/RESULT = {"NAME":"Aoycocr U3S","GPIO":[320,1,321,1,0,2720,0,0,2624,32,2656,224,0,0],"FLAG":0,"BASE":45}
10:32:01.401 MQT: stat/tasmota_0329F7/RESULT = {"Module":{"0":"Aoycocr U3S"}}
10:32:01.656 MQT: stat/tasmota_0329F7/RESULT = {"GPIO0":{"320":"Led_i1"},"GPIO1":{"0":"None"},"GPIO2":{"321":"Led_i2"},"GPIO3":{"0":"None"},"GPIO4":{"0":"None"},"GPIO5":{"2720":"BL0937 CF"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"2624":"HLWBL SEL_i"},"GPIO13":{"32":"Button1"},"GPIO14":{"2656":"HLWBL CF1"},"GPIO15":{"224":"Relay1"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}

```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:
Not using rules

```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
10:32:07.695 CMD: Status 0
10:32:07.705 MQT: stat/tasmota_0329F7/STATUS = {"Status":{"Module":0,"DeviceName":"Aoycocr-6","FriendlyName":["TV-Up Socket"],"Topic":"tasmota_0329F7","ButtonTopic":"0","Power":1,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
10:32:07.737 MQT: stat/tasmota_0329F7/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T16:04:35","StartupUTC":"2021-04-13T17:27:32","Sleep":50,"CfgHolder":4617,"BootCount":831,"BCResetTime":"2020-06-27T16:46:38","SaveCount":2087,"SaveAddress":"F5000"}}
10:32:07.771 MQT: stat/tasmota_0329F7/STATUS2 = {"StatusFWR":{"Version":"9.3.1(tasmota)","BuildDateTime":"2021-03-09T16:12:28","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"437/699"}}
10:32:07.792 MQT: stat/tasmota_0329F7/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["Sorin_AP2y",""],"TelePeriod":300,"Resolution":"568980C0","SetOption":["00008009","2805C8000100060000005A0A000000000000","000002C0","00006000","00000000"]}}
10:32:07.824 MQT: stat/tasmota_0329F7/STATUS4 = {"StatusMEM":{"ProgramSize":593,"Free":408,"Heap":25,"ProgramFlashSize":1024,"FlashSize":1024,"FlashChipId":"144068","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8FDAC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","00000000"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45","Sensors":"1,2,3,4,5,6"}}
10:32:07.857 MQT: stat/tasmota_0329F7/STATUS5 = {"StatusNET":{"Hostname":"tasmota_0329F7-2551","IPAddress":"192.168.10.222","Gateway":"192.168.10.1","Subnetmask":"255.255.255.0","DNSServer":"192.168.10.1","Mac":"C8:2B:96:03:29:F7","Webserver":2,"WifiConfig":2,"WifiPower":17.0}}
10:32:07.880 MQT: stat/tasmota_0329F7/STATUS6 = {"StatusMQT":{"MqttHost":"192.168.10.41","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_0329F7","MqttUser":"tasmota","MqttCount":2,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30}}
10:32:07.903 MQT: stat/tasmota_0329F7/STATUS7 = {"StatusTIM":{"UTC":"2021-04-14T09:32:07","Local":"2021-04-14T10:32:07","StartDST":"2021-03-28T01:00:00","EndDST":"2021-10-31T02:00:00","Timezone":99,"Sunrise":"06:00","Sunset":"19:39"}}
10:32:07.921 MQT: stat/tasmota_0329F7/STATUS9 = {"StatusPTH":{"PowerDelta":[102,0,0],"PowerLow":0,"PowerHigh":0,"VoltageLow":0,"VoltageHigh":0,"CurrentLow":0,"CurrentHigh":0}}
10:32:07.940 MQT: stat/tasmota_0329F7/STATUS10 = {"StatusSNS":{"Time":"2021-04-14T10:32:07","ENERGY":{"TotalStartTime":"2021-03-01T11:31:01","Total":16.65022,"Yesterday":0.45922,"Today":0.13709,"Power":6.8,"ApparentPower":15.6,"ReactivePower":14.1,"Factor":0.44,"Voltage":240,"Current":0.065}}}
10:32:07.964 MQT: stat/tasmota_0329F7/STATUS11 = {"StatusSTS":{"Time":"2021-04-14T10:32:07","Uptime":"0T16:04:35","UptimeSec":57875,"Heap":25,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":2,"POWER":"ON","Wifi":{"AP":1,"SSId":"Sorin_AP2y","BSSId":"2C:4D:54:48:CE:C9","Channel":6,"RSSI":88,"Signal":-56,"LinkCount":1,"Downtime":"0T00:00:09"}}}

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
Not applicable - log provided with the description.

```

### TO REPRODUCE
WattRes 1
PowerDelta 15

### EXPECTED BEHAVIOUR
Expected PowerDelta to take into account the precision of measurement and calculate the delta percentage accurately.

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by isorin** [comment]
I have updated the template.
I can provide more info if needed.

**Comment by arendst** [comment, marked as answer]
As designed.

The energy monitoring hardware is not able to consistently report power useage to the decimal.

The feature is supposed to be used for decent loads anyway.

**Comment by sfromis** [comment]
I'd say that PowerDelta mainly make sense with watt-based deltas, **not** for percents at moderate loads. As a starting point, I'd suggest using `PowerDelta 110` for reasonably reliable operation.

**Reply by isorin** [reply]
![image](https://user-images.githubusercontent.com/17415569/114731668-d6f4db80-9d39-11eb-923d-a165dc973595.png)

It works pretty good for me.

The graphic above is from a "STB+TV" which uses around 6.6-7.0W in standby (as the STB receiver is still powered on in standby) and between 30-45W in use (depending on the ambient light level).

I have a couple of sockets (power calibrated and configured with 1 decimal point precision) and the readings are good enough.
I use one of them to determine if a PC is in sleep or shutdown (the difference is 2-3W in the range of 5W-8W).

I am not trying to use PowerDelta to detect changes in the 6.6-7.0W range, I am simply trying to use "PowerDelta 15" and expect it NOT to trigger for changes in the low end range. I know hardware is not there yet and the precision is not good enough but it will get there.
I also know about PowerDelta 100+ and I know about adding rules to dynamically change the PowerDelta percentage depending on power thresholds but I don't want to complicate things unnecessary.

We have decimal precision for energy measurements, I'm just saying it shouldn't be a big issue to make the PowerDelta calculations with at least 1 decimal point precision.
@arendst If this is something you would consider, I can take a look at the code and see if I can contribute, I think it's still valuable to have the PowerDelta more precise (and not for using it to detect low power deltas, on the contrary, to make it work properly with low power deltas)

**Reply by sfromis** [reply]
To separate ~7W from ~30W, just use a more suitable PowerDelta. Using a watt-based one is better as the percentages works a lot worse at the lower wattage. Due the the genral lack of stability of the energy sensors, it does not make sense to assume that decimals are useful to include in the calculation. Note that the plot you posted above is not the same as what the sensors may deliver every couple of seconds, as the jitter tends to be evened out a lot from telemetry message to the next. However, that jitter may likely affect the PowerDelta calculation, and the log you posted does not show what PowerDelta actually acted on.

I'd suggest just dropping the assumption of a stable sensor providing each reading as an exact value, and simply choose a PowerDelta which does the job, instead of philosophizing about decimals. If you actually want to fuss about numbers, do activate `WebLog 4` to see what happens.

**Reply by isorin** [reply]
Thank you for your input.

The problem is like this:
PowerDelta 15 is the one "that does the job" for me and it's perfect for loads greater than 10W.
The issue is at small loads like the one I posted, around 7W or less, the lack of precision makes it trigger when it shouldn't.
I actually want to avoid these triggers that why I was thinking about increasing the precision of calculation.
Due to the same reason you stated (instability of sensors, jitter, etc.) I think a different approach might be needed.

A setting like "PowerDelta minimum value" (in Watts) to weed out these spikes at a low power range could solve the issue.
Something like 15% delta but no less than 3W (or 5W) would be useful.

I am also considering not using PowerDelta at all and moving the processing into HA.

**Reply by isorin** [reply]
I will try something like this:

```
Backlog PowerDelta 0; Rule1 0; Rule1 5

Rule1
  ON ENERGY#Power>=35 DO Backlog PowerDelta 10; Status 8 BREAK 
  ON ENERGY#Power>=15 DO Backlog PowerDelta 25; Status 8 BREAK 
  ON ENERGY#Power>5 DO Backlog PowerDelta 35; Status 8 BREAK 
  ON ENERGY#Power<=5 DO PowerDelta 100 ENDON

Rule1 1
```

Will this write to flash everytime PowerDelta is set?

**Reply by sfromis** [reply]
Of course, settings are saved to flash when changed. You can use SaveData to control that. And that rule will trigger very often, each sampling cycle, unless you do something to mitigate that.

**Comment by arendst** [comment]
>  If this is something you would consider, I can take a look at the code and see if I can contribute, I think it's still valuable to have the PowerDelta more precise (and not for using it to detect low power deltas, on the contrary, to make it work properly with low power deltas)

I wouldn't waste your time.

**Reply by isorin** [reply]
So basically the code in here:
https://github.com/arendst/Tasmota/blob/4fce3cff7bf5ae3af258ae1c067ead4377e0f354/tasmota/xdrv_03_energy.ino

change "uint16_t" into "float" and "abs" into "fabs" ?
(and power_history into float)

```
  int16_t power_diff[ENERGY_MAX_PHASES] = { 0 };
  for (uint32_t phase = 0; phase < Energy.phase_count; phase++) {
    uint16_t active_power = (uint16_t)(Energy.active_power[phase]);

//    AddLog(LOG_LEVEL_DEBUG, PSTR("NRG: APower %d, HPower0 %d, HPower1 %d, HPower2 %d"), active_power, Energy.power_history[phase][0], Energy.power_history[phase][1], Energy.power_history[phase][2]);

    if (Settings.energy_power_delta[phase]) {
      power_diff[phase] = active_power - Energy.power_history[phase][0];
      uint16_t delta = abs(power_diff[phase]);
      bool threshold_met = false;
      if (delta > 0) {
        if (Settings.energy_power_delta[phase] < 101) {  // 1..100 = Percentage
          uint16_t min_power = (Energy.power_history[phase][0] > active_power) ? active_power : Energy.power_history[phase][0];
          if (0 == min_power) { min_power++; }    // Fix divide by 0 exception (#6741)
          delta = (delta * 100) / min_power;
          if (delta > Settings.energy_power_delta[phase]) {
            threshold_met = true;
          }
        } else {                                  // 101..32000 = Absolute
          if (delta > (Settings.energy_power_delta[phase] -100)) {
            threshold_met = true;
          }
        }
      }
```
